docs/lab/lab04/lab04.md

```diff
@@ -0,0 +1,449 @@
+
+![PWMExample](images/pulse-width-modulation-signal-diagrams-average.png)
+
+:::note
+We can think of the simulated analog signal being directly proportional to the change in digital signal pulse size. The larger the square wave at a given period T, the higher the average analog amplitude output for that period.
+:::
+
+The **duty cycle** of the signal is the percentage of time per period that the signal is high.
+
+![DutyCycle](images/duty_cycle.png)
+
+So if we wanted our led to be at 50% intensity, we would choose a duty cycle of 50%. By quickly switching between high and low, the led appears to the human eye as being at only 50% intensity, when in reality, it's only on at max intensity 50% of the time, and off the rest of the time.
+
+![PWMLed](images/pwm_led.gif)
+
+$$
+
+duty\_cycle = \frac{time\_on}{period} \%
+
+$$
+
+For the RP2040, to generate this PWM signal, a *counter* is used. The PWM counter is controlled by these registers (`X` can be from 0-7, depending on the channel):
+
+- `CHX_CTR` - the actual value of the counter
+- `CHX_CC` - the value that the counter will compare to
+- `CHX_TOP` - the value at which the counter will reset (or *wrap*)
+
+When `CHX_CTR` is reset, the value of the output signal is 1. The counter counts up until it reaches `CHX_CC`, after which the value of the output signal becomes 0. The counter continues to count until it reaches `CHX_TOP`, and then the signal becomes 1 again. This way, by choosing the value of `CHX_CC`, we set the duty cycle of the PWM signal.
+
+![PWMRP2040](images/pwm_rp2040_example.png)
+
+On RP2040, all GPIO pins support PWM. Every two pins share a PWM slice, and each one of them is on a separate channel. 
+
+![RP2040PWMPins](images/pwm_rp2040_pins.png)
+
+:::info
+This means that in order to use a pin as PWM, we need to know what channel it's on, and which output it uses (A or B).
+:::
+
+### Examples of hardware controlled through PWM
+
+- leds
+- motors
+- buzzers
+- RGB leds (what we will be using for this lab)
+
+An **RGB** led is a led that can emit any color, using a combination of red, green and blue light. On the inside, it's actually made up of 3 separate leds:
+- *R* led - to control the intensity of the *red* light
+- *G* led - to control the intensity of the *green* light
+- *B* led - to control the intensity of the *blue* light
+
+By using PWM on the R, G and B leds, we can control each of their intensity to represent any color.
+
+:::info
+For example, if we wanted to create the color purple, we would set the intensity of red and blue to 100%, and the intensity of green to 0%.
+:::
+
+There are two different types of RGB leds:
+
+- common cathode: all LED cathodes are connected together. A LOW signal means off, and a HIGH signal means on at max intensity.
+- common anode: all LED anodes are connected together. A LOW signal means on at max intensity, and a HIGH signal means off.
+
+![CommonAnodeCommonCathode](images/common_anode_common_cathode.png)
```

Comment:
This picture is blurry, is there any way we can sharpen it?

Reply:
Sharpened.